File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        Queue<Integer> queue = new LinkedList<>();
        
        int index = 0;
        int totalWeight = 0;
        for(int i = 0; i < bridge_length; i++){
            queue.add(0); // bridge_length 만큼 queue에 0 add
        }
        
        while(!queue.isEmpty()){
            answer++;
            
            int exited = queue.poll();
            totalWeight -= exited;
            
            if (index < truck_weights.length) {
                if (totalWeight + truck_weights[index] <= weight) {
                    totalWeight += truck_weights[index];
                    queue.add(truck_weights[index++]);
                } else {
                    queue.add(0);
                }
            }
            
        }
        
        
        return answer;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 1
    queue = deque()
    i = 1
    n = len(truck_weights)
    
    queue.append(truck_weights[0])
    current_weight = truck_weights[0] # 현재 다리 위에 있는 총 무게
    
    while queue:        
        if len(queue) == bridge_length:
            out_weight = queue.popleft()
            current_weight -= out_weight
            
        if i >= n and current_weight == 0:
            break
            
        if len(queue) < bridge_length:            
            if i < n and current_weight + truck_weights[i] <= weight:
                queue.append(truck_weights[i])
                current_weight += truck_weights[i]
                i+=1
            else:
                queue.append(0)
        
        answer+=1
    
    return answer + 1